Make Txt base for TextWithCopy component

jmakivic opened the pull request:
<!-- You can remove tags that do not apply. -->
Connects-to: #1465  <!-- waffle convention to track a PR's status through its connected, open issue -->
See: https://www.flowdock.com/app/rulemotion/resin-frontend/threads/mPP--5c357OWJJkaU2ZdUhxa1B5<!-- Refer to any external resource, like a PR, document or discussion -->
Change-type: minor

---
##### Contributor checklist
<!-- For completed items, change [ ] to [x].  -->
- [ ] I have regenerated screenshots for any affected components with `npm run generate-snapshots`
- [ ] I have regenerated jest snapshots for any affected components with `npm run jest -- -u`

##### Reviewer Guidelines
- When submitting a review, please pick:
  - '*Approve*' if this change would be acceptable in the codebase (even if there are minor or cosmetic tweaks that could be improved).
  - '*Request Changes*' if this change would not be acceptable in our codebase (e.g. bugs, changes that will make development harder in future, security/performance issues, etc).
  - '*Comment*' if you don't feel you have enough information to decide either way (e.g. if you have major questions, or you don't understand the context of the change sufficiently to fully review yourself, but want to make a comment)
---

jmakivic commented:
![Truncate](https://user-images.githubusercontent.com/8771822/149014190-37a2f2fd-f13a-4019-a616-92a0851c2ceb.png)

I added some properties to Txt which make it look like highlighted code. I also added properties which truncate the content it if it gets too long. I wrapped Txt in the <Copy> component. This screenshot features DeviceSummaryCard.tsx.

I applied these changes to  the public IP address and the MAC address. As you can see, the long public IP address gets truncated if there is not enough space. 

This is what the code looks like  on DeviceSummaryCard.tsx . 
```
{device.public_address && (
	<div>
	   <Label>{t('labels.public_ip_address')}</Label>
	   <Copy content={device.public_address} onClick={function noRefCheck(){}} title={device.public_address} show="always" >
	              <Txt style={{width: "80%", whiteSpace: "nowrap", display:"inline-block" }} code truncate>{device.public_address</Txt>
	   </Copy>
	</div>
)}
{device?.mac_address != null && (
	<div>
	   <Label>{t('labels.mac_address')}</Label>
	   <Copy content={device.mac_address}  onClick={function noRefCheck(){}} show="always" title={device.mac_address} >
			<Txt style={{ whiteSpace: "nowrap", display:"inline-block" }} code truncate>{device.mac_address}</Txt>
	   </Copy>
	 </div>
)}
```

thgreasi commented on src/components/Txt/index.tsx:
```diff
@@ -3,7 +3,23 @@ import styled, { css } from 'styled-components';
+		? css`
+				font-family: ${(props) => props.theme.monospace};
+				padding: 2px 4px;
+				font-size: 90%;
```
This get's overrides by the `font-size: 1em;` right below.

thgreasi commented on src/components/Txt/index.tsx:
```diff
@@ -3,7 +3,23 @@ import styled, { css } from 'styled-components';
+				border-radius: 2px;
+				white-space: normal;
+				word-wrap: break-word;
+				font-size: 1em;
```
Let's remove this, since it will not have any effect after we remove the `font-size: 90%;` from above.

thgreasi commented on src/components/Copy/index.tsx:
```diff
@@ -14,6 +14,9 @@ const ClipboardContainer = styled(Box)`
 	font-size: 0.875em;
 	line-height: ${(props) =>
 		px(props.theme.lineHeight * props.theme.fontSizes[2])};
+	&:hover {
+		visibility: visible;
+	}
```
I checked out the respective UI branch and couldn't see any difference w/ or w/o this line.
Can you expand why we needed this?

jmakivic replied:
I was trying to replicate the properties of the <Copy> component onto the `copy` prop. It turns out this element isn't necessary for the `copy` prop. I removed it.

thgreasi commented:
Btw it's fine to use `Resolves: #<number>` in Rendition & in the UI in case the PR completely fixes the issue and it can be closed on merge.

jmakivic commented on src/components/Txt/index.tsx:
```diff
@@ -119,6 +134,8 @@ Base.displayName = 'Txt';
 Base.span = Factory('span');
 Base.p = Factory('p');
 
+// defining the stuff for copy, maybe this can be it's own thing in another
```
That was a comment which I wrote for myself which I forgot to delete. I removed it.

jmakivic replied:
This was just a comment I left for myself which I forgot to delete. It's gone.

jmakivic commented on src/components/Txt/index.tsx:
```diff
@@ -107,6 +121,7 @@ export interface InternalTxtProps extends React.HTMLAttributes<HTMLElement> {
 	align?: Align;
 	/** If true, replace the text not contained in the container with three dots */
 	truncate?: boolean;
+	code?: boolean;
```
I moved it to be after the italic prop.

jmakivic replied:
I moved it after `italic`.

jmakivic commented on src/components/Copy/index.tsx:
```diff
@@ -22,7 +25,7 @@ const CopyContainer = styled(Flex)<{
 		display: ${(props) => (props.show === 'always' ? 'inline' : 'none')};
 	}
-
+	}
```
Do you mean the extra semicolon or the code on line 26? Because the code on line 26 seems valid to me.

jmakivic commented on package.json:
```diff
@@ -136,6 +136,7 @@
     "remark-breaks": "^2.0.1",
     "remark-parse": "^8.0.3",
     "remark-rehype": "^7.0.0",
+    "rendition": "^25.2.0",
```
I undid these changes to the package.json.

jmakivic commented on src/components/Txt/index.tsx:
```diff
@@ -93,7 +108,6 @@ export type Align =
 	| 'unset';
 
 export interface InternalTxtProps extends React.HTMLAttributes<HTMLElement> {
-	/** If true, render text in a monospace font */
```
I added it back in. Must have gotten misplaced while I was editing.

JSReds commented on src/components/Copy/index.tsx:
```diff
@@ -47,6 +46,7 @@ const BaseCopy = ({
 			{...otherProps}
 		>
 			{children}
+
```
we can probably just revert the entire fille unless this is a change made by the linter

JSReds commented on src/components/Txt/index.tsx:
```diff
@@ -48,24 +60,54 @@ export const align = style({
 	});
 };
 
+/*const Factory = (tag?: string) => {
+	return asRendition<React.FunctionComponent<TxtProps>>((props: any) => {
+		return <BaseTxt as={tag} {...props}>
+    	{props.children}
+    	{!!props.copy &&
+        	<Copy content={props.children} show={props.copy} {...(props.copy.show && { showCopyButton: props.copy.show })}/>
+    	}
+		</BaseTxt>;
+		
+	});
+};*/
```
let's remove this if not needed

JSReds commented on src/extra/MarkdownEditor/defaultStyle.ts:
```diff
@@ -207,16 +207,16 @@ export default css`
 	}
 	.CodeMirror-scroll {
 		overflow: scroll !important;
-		margin-bottom: -30px;
```
Just to make sure, how this file was changed ?

jmakivic replied:
I'm not sure. I don't recall changing it.

myarmolinsky replied:
If this was not intentional then please revert it

myarmolinsky replied:
Please revert the changes to this file then

JSReds commented on src/components/Txt/index.tsx:
```diff
@@ -107,6 +151,10 @@ export interface InternalTxtProps extends React.HTMLAttributes<HTMLElement> {
+	/** If true, it adds the copy icon and functionality at the end of a Txt component */
+	copy?: boolean;
+	/**  If true, the copy icon only appears when hovering over the content to be copied */
+	hoverCopy?: boolean;
```
I would honestly keep it as `showCopyButton` maybe we can change it to `showCopyMode`, just in case we want to handle multiple states, and also to keep the same property we have in the Copy component.

```suggestion
	showCopyMode?: 'always' | 'hover';
```

JSReds commented on src/components/Txt/index.tsx:
```diff
@@ -48,24 +60,54 @@ export const align = style({
+		return props.copy ? (
+			<Copy
+				content={props.children}
+				show={props.hoverCopy ? 'hover' : 'always'}
```
if we keep the same property as suggested in another comment

```suggestion
				show={props.showCopyMode || 'always'}
```

JSReds commented on src/components/TextWithCopy/index.tsx:
```diff
@@ -12,7 +12,6 @@ import { Txt, TxtProps } from '../Txt';
 const Wrapper = styled(Txt.span)<InternalTextWithCopyProps>`
```
I suppose that once we add the property to the Txt component, this file can be deleted

JSReds commented on src/components/Txt/index.tsx:
```diff
@@ -48,24 +60,54 @@ export const align = style({
 	cssProperty: 'text-align',
 });
 
+export const Lopy = (props: ThemedTxtProps) =>
+	props.truncate &&
+	css`
+		cursor: pointer;
+	`;
+
+export const LopyWrapper = styled.span`
+	display: inline-block;
+`;
+
```
What is Lopy ? and where are we using them ?

JSReds commented on src/components/Txt/index.tsx:
```diff
@@ -48,24 +60,54 @@ export const align = style({
-		return <BaseTxt as={tag} {...props} />;
+		return props.copy ? (
+			<Copy
+				content={props.children}
```
Not sure why TS is not complaining here but you are passing a possible JSX element as a string property.  This is not allowed, or at least this is not correct. You should pass porps.copy that should be a string containing what we want to copy

jmakivic replied:
I understand. I think this is what Matt brought up in the flowdock discussion. I changed content to be props.copy.

jmakivic commented on src/components/Txt/index.tsx:
```diff
@@ -107,6 +151,10 @@ export interface InternalTxtProps extends React.HTMLAttributes<HTMLElement> {
 	/** If true, replace the text not contained in the container with three dots */
 	truncate?: boolean;
+	/** If true, it adds the copy icon and functionality at the end of a Txt component */
+	copy?: boolean;
```
Changed it.

myarmolinsky commented on src/components/Txt/index.tsx:
```diff
@@ -52,17 +64,22 @@ const BaseTxt = styled.div<TxtProps>`
 	return asRendition<React.FunctionComponent<TxtProps>>((props: any) => {
-		return <BaseTxt as={tag} {...props} />;
+		return props.copy ? (
+			<Copy content={props.copy} show={props.showCopyMode || 'always'}>
+				<BaseTxt as={tag} {...props} />
+			</Copy>
```
Since `Copy` isn't receiving `{...props}`, its styling will be off compared to `BaseTxt`'s. So, I'm not certain but I think `{...props}` should be passed to `Copy` instead of `BaseTxt`. I believe `BaseTxt` will not be happy about no longer receiving `{...props}` so please look into that. If `BaseTxt` requires `props` to function, maybe there is a specific property in `props` that we can destructure and pass to it in order to get it to work?

Giving `Copy` `{...props}` won't be enough to get its styling working properly. I was testing by updating the `Txt.span` containing the `app.total_devices` in the UI's `ApplicationListBox` component so you can use that as your test case. You'll see that while giving `Copy` `{...props}` makes it resize and obey color and other styling, it does not center the text like it is supposed to. Please see if you can find what's causing that and how to fix it.

myarmolinsky commented on src/components/Txt/index.tsx:
```diff
@@ -107,6 +126,10 @@ export interface InternalTxtProps extends React.HTMLAttributes<HTMLElement> {
 	truncate?: boolean;
+	/** Adds copy functionality to Txt and accepts the value to be copied as a string */
+	copy?: string;
+	/**  Setting to 'always' ensures that the copy icon is always visible. 'hover' ensures that the copy icon is only visible on hover. If this prop is not included, the copy icon is set to hover */
```
```suggestion
	/**  Setting to 'always' ensures that the copy icon is always visible. 'hover' ensures that the copy icon is only visible on hover. By default, this is set to 'hover'. */
```
```suggestion
	/**  Setting to 'always' ensures that the copy icon is always visible. 'hover' ensures that the copy icon is only visible on hover. If this prop is not included, the copy icon is set to hover */
```

myarmolinsky commented on src/components/Txt/index.tsx:
```diff
@@ -107,6 +126,10 @@ export interface InternalTxtProps extends React.HTMLAttributes<HTMLElement> {
 	align?: Align;
 	/** If true, replace the text not contained in the container with three dots */
 	truncate?: boolean;
+	/** Adds copy functionality to Txt and accepts the value to be copied as a string */
```
```suggestion
	/** Displays a copy icon which copies the given string to the user's clipboard onClick */
```

myarmolinsky commented on src/components/Copy/index.tsx:
```diff
@@ -22,7 +22,6 @@ const CopyContainer = styled(Flex)<{
 	${ClipboardContainer} {
 		display: ${(props) => (props.show === 'always' ? 'inline' : 'none')};
 	}
-
```
Please revert this file as it is unchanged

myarmolinsky commented on src/components/Txt/index.tsx:
```diff
@@ -107,6 +155,12 @@ export interface InternalTxtProps extends React.HTMLAttributes<HTMLElement> {
+	copy?: string;
+	/**  Setting to 'always' ensures that the copy icon is always visible. 'hover' ensures that the copy icon is only visible on hover. By default, this is set to 'hover'. */
+	showCopyMode?: 'hover' | 'always';
+	/** Displays Txt as inline-block and vertical aligns items at the bottom. Good for aligning text with icons. */
+	iconalign?: boolean;
```
Why do we need this as a property? The Txt should use the appropriate display by default. If we want the user to be able to change it, we should have a property called `display` (it exists on both BoxProps and FlexProps). Alignment should be a separate property as well, likely and preferably from FlexProps or BoxProps, and it should have a good, sensible default value

JSReds replied:
Agree with this. We could probably remove it without any replacement, unless I'm forgetting any strange cases

myarmolinsky commented on src/components/Txt/index.tsx:
```diff
@@ -48,21 +63,52 @@ export const align = style({
 	return asRendition<React.FunctionComponent<TxtProps>>((props: any) => {
-		return <BaseTxt as={tag} {...props} />;
+		return props.copy ? (
+			<div>
+				<BaseTxt as={tag} {...props} iconalign />
+				<Copy
+					tooltip={{ text: 'Copied!', trigger: 'click' }}
+					onClick={() => copyToClipboard(props.copy)}
+					showCopyButton={props.showCopyMode || 'always'}
+					{...props}
+				>
+					<FontAwesomeIcon icon={faCopy} />
+				</Copy>
+			</div>
```
Why do we need to wrap this in a div? And howcome Copy is wrapping a copy icon instead of the BaseTxt? If the user provides copy, the result should be something like `<Copy><BaseTxt /></Copy>`. Any styling updates that need to be made in order for it to work properly in any contexts (consider the ones in our UI) should be applied to one of those two components.

JSReds replied:
@myarmolinsky IMHO this is the correct structure. Copy as it was before is just a styled component that wrap an icon, as for the style we already discussed that the icon should not have any other than the font size. (ref: https://www.flowdock.com/app/rulemotion/f-rendition/threads/tcqn5E_rvx-y9M66LSJHg57t3Ki)

Also @jmakivic what about:

```
			<div>
				<BaseTxt as={tag} {...props} iconalign />
{ props.copy && (
				<Copy
					tooltip={{ text: 'Copied!', trigger: 'click' }}
					onClick={() => copyToClipboard(props.copy)}
					showCopyButton={props.showCopyMode || 'always'}
          fontSize={props.fontSize}
				>
					<FontAwesomeIcon icon={faCopy} />
				</Copy> )
			}
			</div>
```

JSReds commented on src/components/Txt/index.tsx:
```diff
@@ -48,21 +63,52 @@ export const align = style({
+	${iconalign}
 `;
-
+// style={{ display: 'inline-block', verticalAlign: 'bottom' }}
```
Remember to remove

JSReds commented on src/components/Txt/index.tsx:
```diff
@@ -48,22 +64,63 @@ export const align = style({
+`;
+
+const copy = (props: ThemedTxtProps) =>
+	props.copy &&
```
Let's make this check a boolean

```suggestion
	!!props.cop &&
```

JSReds reviewed (commented):
Other than that, LGTM :)

jmakivic commented:
I changed the `onCopy` prop back to `onCopyClick`. `onCopy` appears to be an existing react property. I saw this error when I added it as a Txt prop:

src/components/Link/index.tsx:12:4 - error TS2345: Argument of type '(props: ThemedTxtProps) => FlattenSimpleInterpolation | null' is not assignable to parameter of type 'Interpolation<ThemedStyledProps<Pick<DetailedHTMLProps<AnchorHTMLAttributes<HTMLAnchorElement>, HTMLAnchorElement>, "slot" | ... 263 more ... | "referrerPolicy"> & { ...; } & InternalLinkProps, any>>'.
  Type '(props: ThemedTxtProps) => FlattenSimpleInterpolation | null' is not assignable to type 'InterpolationFunction<ThemedStyledProps<Pick<DetailedHTMLProps<AnchorHTMLAttributes<HTMLAnchorElement>, HTMLAnchorElement>, "slot" | ... 263 more ... | "referrerPolicy"> & { ...; } & InternalLinkProps, any>>'.
    Types of parameters 'props' and 'props' are incompatible.
      Type 'ThemedStyledProps<Pick<DetailedHTMLProps<AnchorHTMLAttributes<HTMLAnchorElement>, HTMLAnchorElement>, "slot" | ... 263 more ... | "referrerPolicy"> & { ...; } & InternalLinkProps, any>' is not assignable to type 'ThemedTxtProps'.
        Types of property 'onCopy' are incompatible.
          Type '(((event: ClipboardEvent<HTMLAnchorElement>) => void) & ((event: ClipboardEvent<HTMLElement>) => void)) | undefined' is not assignable to type '(() => void) | undefined'.
            Type '((event: ClipboardEvent<HTMLAnchorElement>) => void) & ((event: ClipboardEvent<HTMLElement>) => void)' is not assignable to type '(() => void) | undefined'.
              Type '((event: ClipboardEvent<HTMLAnchorElement>) => void) & ((event: ClipboardEvent<HTMLElement>) => void)' is not assignable to type '() => void'.

This problem was resolved when I renamed `onCopy` to `onCopyClick`.
